File: app/src/main/java/ru/e6atb/chat/VoiceCall.java
package ru.e6atb.chat;

import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;
import android.media.AudioFormat;
import android.media.AudioManager;
import android.media.AudioRecord;
import android.media.AudioTrack;
import android.media.MediaRecorder;

import java.io.Closeable;
import java.io.IOException;

import rs.ove.crypt.proto.SecureSessionV4;

final class VoiceCall {
	static final String STATE_ENDED = "call_ended";
	static final String STATE_CONNECTED = "voice_connected";
	static final String STATE_CONNECTION_CLOSED = "voice_connection_closed";
	static final String STATE_MICROPHONE_PERMISSION_DENIED = "voice_microphone_permission_denied";
	static final String STATE_ERROR_PREFIX = "voice_error:";
	static final String STATE_SEND_ERROR_PREFIX = "voice_send_error:";

	private static final int SAMPLE_RATE = 8000;
	private static final int FRAME_BYTES = 640;

	private volatile boolean running;
	private SimpleWebSocket ws;
	private SecureSessionV4 crypt;
	private AudioRecord recorder;
	private AudioTrack player;
	private Thread readThread;
	private Thread micThread;
	private Listener listener;

	interface Listener {
		void onState(String text);
	}

	boolean running() {
		return running;
	}

	void start(final Context context, final String url, Listener listener) {
		if (running) {
			return;
		}
		this.listener = listener;
		running = true;
		readThread = new Thread(new Runnable() {
			@Override
			public void run() {
				VoiceCall.this.run(context == null ? null : context.getApplicationContext(), url);
			}
		}, "e6atb-voice-read");
		readThread.start();
	}

	void stop() {
		stop(true);
	}

	private void stop(boolean notify) {
		running = false;
		crypt = null;
		closeQuietly(ws);
		ws = null;
		releaseRecorder();
		releasePlayer();
		if (notify) state(STATE_ENDED);
	}

	private void run(Context context, String url) {
		boolean remoteClosed = false;
		try {
			ws = new SimpleWebSocket();
			ws.connect(url);
			crypt = connectCrypt(context, ws);
			setupPlayer(context);
			setupRecorder(context);
			player.play();
			recorder.startRecording();
			startMic();
			state(STATE_CONNECTED);

			while (running) {
				SimpleWebSocket.Frame frame = ws.readFrame();
				if (frame.opcode == SimpleWebSocket.CLOSE) {
					remoteClosed = true;
					break;
				}
				if (frame.opcode == SimpleWebSocket.BINARY && frame.payload.length > 0) {
					byte[] plain = crypt.openApplicationRecord(frame.payload);
					if (plain.length > 0) {
						player.write(plain, 0, plain.length);
					}
				}
			}
		} catch (SecurityException e) {
			state(STATE_MICROPHONE_PERMISSION_DENIED);
		} catch (IOException e) {
			if (running) {
				state(STATE_ERROR_PREFIX + errorText(e));
			}
		} catch (Exception e) {
			if (running) {
				state(STATE_ERROR_PREFIX + errorText(e));
			}
		} finally {
			if (running && remoteClosed) {
				state(STATE_CONNECTION_CLOSED);
			}
			stop(false);
		}
	}

	private SecureSessionV4 connectCrypt(Context context, SimpleWebSocket ws) throws Exception {
		SecureSessionV4.ClientHello hello = SecureSessionV4.createClientHello();
		byte[] message = hello.message();
		ws.sendBinary(message, message.length);
		SimpleWebSocket.Frame frame = ws.readFrame();
		if (frame.opcode != SimpleWebSocket.BINARY) {
			throw new IOException(text(context, R.string.status_crypto_handshake_failed));
		}
		return SecureSessionV4.openClient(hello, frame.payload);
	}

	private void setupPlayer(Context context) {
		int min = AudioTrack.getMinBufferSize(
				SAMPLE_RATE,
				AudioFormat.CHANNEL_OUT_MONO,
				AudioFormat.ENCODING_PCM_16BIT
		);
		int size = Math.max(min, FRAME_BYTES * 8);
		player = new AudioTrack(
				AudioManager.STREAM_VOICE_CALL,
				SAMPLE_RATE,
				AudioFormat.CHANNEL_OUT_MONO,
				AudioFormat.ENCODING_PCM_16BIT,
				size,
				AudioTrack.MODE_STREAM
		);
		if (player.getState() != AudioTrack.STATE_INITIALIZED) {
			throw new IllegalStateException(text(context, R.string.status_speaker_unavailable));
		}
	}

	private void setupRecorder(Context context) {
		if (context != null
				&& context.checkCallingOrSelfPermission(Manifest.permission.RECORD_AUDIO)
				!= PackageManager.PERMISSION_GRANTED) {
			throw new SecurityException(text(context, R.string.status_microphone_denied));
		}
		int min = AudioRecord.getMinBufferSize(
				SAMPLE_RATE,
				AudioFormat.CHANNEL_IN_MONO,
				AudioFormat.ENCODING_PCM_16BIT
		);
		int size = Math.max(min, FRAME_BYTES * 8);
		int source = MediaRecorder.AudioSource.MIC;
		if (android.os.Build.VERSION.SDK_INT >= 11) {
			source = audioSource("VOICE_COMMUNICATION", source);
		}
		recorder = new AudioRecord(
				source,
				SAMPLE_RATE,
				AudioFormat.CHANNEL_IN_MONO,
				AudioFormat.ENCODING_PCM_16BIT,
				size
		);
		if (recorder.getState() != AudioRecord.STATE_INITIALIZED) {
			throw new IllegalStateException(text(context, R.string.status_microphone_unavailable));
		}
	}

	private int audioSource(String name, int fallback) {
		try {
			return MediaRecorder.AudioSource.class.getField(name).getInt(null);
		} catch (Exception ignored) {
			return fallback;
		}
	}

	private void startMic() {
		micThread = new Thread(new Runnable() {
			@Override
			public void run() {
				byte[] buf = new byte[FRAME_BYTES];
				while (running) {
					int n = recorder.read(buf, 0, buf.length);
					if (n > 0) {
						try {
							byte[] plain = new byte[n];
							System.arraycopy(buf, 0, plain, 0, n);
							byte[] sealed = crypt.sealApplicationRecord(plain);
							ws.sendBinary(sealed, sealed.length);
						} catch (Exception e) {
							if (running) {
								state(STATE_SEND_ERROR_PREFIX + errorText(e));
							}
							running = false;
						}
					}
				}
			}
		}, "e6atb-voice-mic");
		micThread.start();
	}

	private void releaseRecorder() {
		AudioRecord r = recorder;
		recorder = null;
		if (r == null) {
			return;
		}
		try {
			r.stop();
		} catch (Exception ignored) {
		}
		r.release();
	}

	private void releasePlayer() {
		AudioTrack p = player;
		player = null;
		if (p == null) {
			return;
		}
		try {
			p.stop();
		} catch (Exception ignored) {
		}
		p.release();
	}

	private void state(String text) {
		Listener l = listener;
		if (l != null) {
			l.onState(text);
		}
	}

	private static void closeQuietly(Closeable closeable) {
		if (closeable == null) {
			return;
		}
		try {
			closeable.close();
		} catch (Exception ignored) {
		}
	}

	private static String errorText(Throwable error) {
		if (error == null) return "";
		String message = error.getMessage();
		if (message == null || message.trim().length() == 0) {
			message = error.getClass().getSimpleName();
		}
		return message;
	}

	private static String text(Context context, int resId) {
		return context == null ? "" : context.getString(resId);
	}
}

File: app/src/main/kotlin/ru/e6atb/chat/VoiceCall.kt
package ru.e6atb.chat

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.media.AudioFormat
import android.media.AudioManager
import android.media.AudioRecord
import android.media.AudioTrack
import android.media.MediaRecorder
import rs.ove.crypt.proto.NativeMst5
import java.io.IOException
import java.util.Locale

internal class VoiceCall {
    interface Listener { fun onState(text: String?) }
    @Volatile private var active = false; @Volatile private var stream = 0L; private var recorder: AudioRecord? = null; private var player: AudioTrack? = null; private var readThread: Thread? = null; private var micThread: Thread? = null; private var listener: Listener? = null
    fun running(): Boolean = active
    fun start(context: Context?, access: MST5.VoiceAccess?, listener: Listener?) { if (active) return; this.listener = listener; active = true; readThread = Thread({ run(context?.applicationContext, access) }, "e6atb-voice-read").also { it.start() } }
    fun stop() = stop(true)
    private fun stop(notify: Boolean) { active = false; val current = stream; stream = 0; NativeMst5.closeVoice(current); releaseRecorder(); releasePlayer(); if (notify) state(STATE_ENDED) }
    private fun run(context: Context?, access: MST5.VoiceAccess?) {
        var remoteClosed = false
        try {
            val current = access ?: throw IOException("voice access is unavailable"); stream = NativeMst5.openVoice(current.endpoint, current.serverPublicKey, current.ticket); setupPlayer(context); setupRecorder(context); player!!.play(); recorder!!.startRecording(); startMic(); state(STATE_CONNECTED)
            while (active) { val pcm = NativeMst5.receiveVoice(stream); if (pcm.isNotEmpty()) player?.write(pcm, 0, pcm.size) }
        } catch (_: SecurityException) { state(STATE_MICROPHONE_PERMISSION_DENIED) }
        catch (error: IOException) { if (active) { remoteClosed = connectionClosed(error); if (!remoteClosed) state(STATE_ERROR_PREFIX + errorText(error)) } }
        catch (error: Exception) { if (active) state(STATE_ERROR_PREFIX + errorText(error)) }
        finally { if (active && remoteClosed) state(STATE_CONNECTION_CLOSED); stop(false) }
    }
    private fun setupPlayer(context: Context?) { val size = maxOf(AudioTrack.getMinBufferSize(SAMPLE_RATE, AudioFormat.CHANNEL_OUT_MONO, AudioFormat.ENCODING_PCM_16BIT), FRAME_BYTES * 8); player = AudioTrack(AudioManager.STREAM_VOICE_CALL, SAMPLE_RATE, AudioFormat.CHANNEL_OUT_MONO, AudioFormat.ENCODING_PCM_16BIT, size, AudioTrack.MODE_STREAM); if (player!!.state != AudioTrack.STATE_INITIALIZED) throw IllegalStateException(text(context, R.string.status_speaker_unavailable)) }
    private fun setupRecorder(context: Context?) { if (context != null && context.checkCallingOrSelfPermission(Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) throw SecurityException(text(context, R.string.status_microphone_denied)); val size = maxOf(AudioRecord.getMinBufferSize(SAMPLE_RATE, AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT), FRAME_BYTES * 8); val source = if (android.os.Build.VERSION.SDK_INT >= 11) audioSource("VOICE_COMMUNICATION", MediaRecorder.AudioSource.MIC) else MediaRecorder.AudioSource.MIC; recorder = AudioRecord(source, SAMPLE_RATE, AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT, size); if (recorder!!.state != AudioRecord.STATE_INITIALIZED) throw IllegalStateException(text(context, R.string.status_microphone_unavailable)) }
    private fun audioSource(name: String, fallback: Int): Int = try { MediaRecorder.AudioSource::class.java.getField(name).getInt(null) } catch (_: Exception) { fallback }
    private fun startMic() { micThread = Thread({ val buffer = ByteArray(FRAME_BYTES); while (active) { val count = recorder?.read(buffer, 0, buffer.size) ?: -1; if (count > 0) try { NativeMst5.sendVoice(stream, buffer.copyOf(count)) } catch (error: Exception) { if (active) state(STATE_SEND_ERROR_PREFIX + errorText(error)); active = false } } }, "e6atb-voice-mic").also { it.start() } }
    private fun releaseRecorder() { recorder?.let { try { it.stop() } catch (_: Exception) {}; it.release() }; recorder = null }
    private fun releasePlayer() { player?.let { try { it.stop() } catch (_: Exception) {}; it.release() }; player = null }
    private fun state(text: String) { listener?.onState(text) }
    private fun errorText(error: Throwable?): String { if (error == null) return ""; return error.message?.takeIf { it.trim().isNotEmpty() } ?: error.javaClass.simpleName }
    private fun connectionClosed(error: Throwable): Boolean = errorText(error).lowercase(Locale.US).let { it.contains("stream closed") || it.contains("connection closed") || it.contains("broken pipe") }
    private fun text(context: Context?, resId: Int): String = context?.getString(resId) ?: ""
    companion object { @JvmField val STATE_ENDED = "call_ended"; @JvmField val STATE_CONNECTED = "voice_connected"; @JvmField val STATE_CONNECTION_CLOSED = "voice_connection_closed"; @JvmField val STATE_MICROPHONE_PERMISSION_DENIED = "voice_microphone_permission_denied"; @JvmField val STATE_ERROR_PREFIX = "voice_error:"; @JvmField val STATE_SEND_ERROR_PREFIX = "voice_send_error:"; private const val SAMPLE_RATE = 8000; private const val FRAME_BYTES = 640 }
}
